element_zstack/bossdb.py

```diff
@@ -0,0 +1,226 @@
+        if upload_type == "image":
+            return "https://neuroglancer.bossdb.io/#!" + str(
+                {
+                    "layers": {
+                        "type": f"{type}",
+                        "source": base_url,
+                        "tab": "source",
+                        "name": f"{experiment}",
+                    }
+                }
+            )
+        elif upload_type == "annotation":
+            return "https://neuroglancer.bossdb.io/#!" + str(
+                {
+                    "layers": {
+                        "type": "segmentation",
+                        "source": base_url + "-seg",
+                        "tab": "annotations",
+                        "name": f"{channel}-seg",
+                    }
+                }
+            )
+        elif upload_type == "image+annotation":
+            return "https://neuroglancer.bossdb.io/#!" + str(
+                {
+                    "layers": [
+                        {
+                            "type": "image",
+                            "source": base_url,
+                            "tab": "source",
+                            "name": f"{experiment}",
+                        },
+                        {
+                            "type": "segmentation",
+                            "source": base_url + "-seg",
+                            "tab": "annotations",
+                            "name": f"{channel}-seg",
+                        },
+                    ]
+                }
+            )
+
+    def make(self, key):
+        """Upload data to bossdb."""
+
+        collection, experiment, channel = (VolumeUploadTask & key).fetch1(
+            "collection_name", "experiment_name", "channel_name"
+        )
+
+        voxel_width, voxel_height, voxel_depth = (volume.VoxelSize & key).fetch1(
+            "width", "height", "depth"
+        )
+
+        description = ["image", "annotation", "image+annotation"]
+        full_data = []
+        boss_url = []
+        neuroglancer_url = []
+
+        full_data.append((volume.Volume & key).fetch1("volume"))
+        boss_url.append(f"bossdb://{collection}/{experiment}/{channel}")
+
+        z_size, y_size, x_size = (volume.Volume & key).fetch1(
+            "px_depth", "px_height", "px_width"
+        )
+        segmentation_data = np.zeros((z_size, y_size, x_size))
+
+        mask_ids, x_mask_pix, y_mask_pix, z_mask_pix = (
+            volume.Segmentation.Mask & key
+        ).fetch("mask", "mask_xpix", "mask_ypix", "mask_zpix")
+
+        for idx, mask in enumerate(mask_ids):
+            segmentation_data[
+                np.s_[z_mask_pix[idx], y_mask_pix[idx], x_mask_pix[idx]]
+            ] = mask
+        full_data.append(segmentation_data.astype("uint64"))
+
+        boss_url.append(f"bossdb://{collection}/{experiment}/{channel}-seg")
+
+        for url, data, desc in zip(boss_url, full_data, description):
```

Comment:
I think I am misunderstanding, but `description` has 3 items, and `boss_url` and `full_data` only have 2 items.